### PR TITLE
Codechange: Rearrange (code) of AI/GS debug window.

### PR DESCRIPTION
**Pull request by PeterN**
## Motivation / Problem

The AI/GS debug window updated its state as it was drawn, and would redraw again if some state had changed.

Many widgets were drawn in OnPaint() instead of DrawWidget(), and the company icon was drawn incorrect based on unscaled hardcoded pixel dimensions.

image

## Description

Instead, update state either during OnInvalidateData or before any drawing commences, and draw everything through DrawWidget() handler, via smaller class methods to break it up.

Additionally the "Game Script" button is moved next to the Company/AI selector which makes it more obvious that it is part of the same exclusive toggle.

image

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, game_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/script/script_gui.cpp [hunk 1]
```diff
@@ -929,6 +895,76 @@ struct ScriptDebugWindow : public Window {
...
+
+			NWidgetCore *button = this->GetWidget<NWidgetCore>(i + WID_SCRD_COMPANY_BUTTON_START);
+			button->SetDisabled(!valid);
+			button->SetLowered(script_debug_company == i);
+			SetScriptButtonColour(*button, dead, paused);
+		}
+	}
+
+	/**
+	 * Update state of game scrript button.
```
Comment (2TallTyler) [inline src/script/script_gui.cpp]:
```suggestion
	 * Update state of game script button.
```

File: src/script/script_gui.cpp [hunk 2]
```diff
@@ -929,6 +895,76 @@ struct ScriptDebugWindow : public Window {
...
+		if (this->vscroll->GetCount() != scroll_count) {
+			this->vscroll->SetCount(scroll_count);
+
+			/* We need a repaint */
+			this->SetWidgetDirty(WID_SCRD_SCROLLBAR);
+		}
+
+		if (log.empty()) return;
+
+		/* Detect when the user scrolls the window. Enable autoscroll when the
```
Comment (2TallTyler) [inline src/script/script_gui.cpp]: I know you just moved stuff around, but this comment could probably be on one line.

File: src/script/script_gui.cpp [hunk 3]
```diff
@@ -896,9 +830,41 @@ struct ScriptDebugWindow : public Window {
...
+		}
+	}
+
+	/**
+	 * Draw a company button icon.
+	 * @param r Rect area to draw within.
+	 * @param widget Widget index to start.
+	 * @param start Widget index of first company button.
+	 */
+	void DrawWidgetCompany(const Rect &r, int widget, int start) const
```
Comment (2TallTyler) [inline src/script/script_gui.cpp]:
```suggestion
	void DrawWidgetCompanyButton(const Rect &r, int widget, int start) const
```

Maybe? To make it clear it's not the log of that AI, and just the button.

**Review by 2TallTyler — COMMENTED** [review]
The padding around the company and script buttons looks strange to me -- it's all very busy and crammed up against the bar below containing the script name, the settings button, and the reload button. I wonder if increasing (or maybe eliminating) said padding would help with the visual confusion. 

I know that's out of scope, but since you're there anyway... (Feel free to say "another PR" or "do it yourself" 😛)

**Comment by PeterN** [comment]
> The padding around the company and script buttons looks strange to me -- it's all very busy and crammed up against the bar below containing the script name, the settings button, and the reload button. I wonder if increasing (or maybe eliminating) said padding would help with the visual confusion.
> 
> I know that's out of scope, but since you're there anyway... (Feel free to say "another PR" or "do it yourself" 😛)

I did adjust it from uneven padding where it overlaps or touches borders, so the style used on the Detailed Performance Rating window.

The only other place with this style of list is the Cargo Flow Legend window, which still has touching borders (because I've not touched it yet) so... dunno.

Maybe the button row below could go above the company buttons, or at the bottom with the other buttons. Hmm.

**Comment by glx22** [comment]
Main issue in this window is the number of "tabs"

Maybe the 2 rows of 8 companies (with the empty space because there are only 15) could be replaced with 3 rows of 5.
At least it could allow to pack company buttons and game script button, giving some space on their right which could be filled with a frame for script name and and the 2 buttons.

Another option to keep the tab style is to put tabs in a single column on the left.

It's very hard for this window to look "nice".

**Comment by PeterN** [comment]
All layout changes have been undone, as it seems people would rather have a full redesign than a minor tweak.

**Review by 2TallTyler — APPROVED** [review]
Sorry, I wasn't trying to argue that your layout change wasn't an improvement. 😦 

This might need a new commit title when merging, because it's now a codechange rather than a change. Maybe `Codechange: Refactor AI/GS debug window`.

**Pull request merged**